src/backend/planner/seq_scan_plan.cpp

```diff
@@ -24,89 +24,89 @@ namespace planner {
-        oid_t col_id = GetColumnIds()[it];
-        output.WriteInt(static_cast<int>(col_id));
-    }
+bool SeqScanPlan::SerializeTo(SerializeOutput &output) const {
+
+  // A placeholder for the total size written at the end
+  int start = output.Position();
+  output.WriteInt(-1);
+
+  // Write the SeqScanPlan type
+  PlanNodeType plan_type = GetPlanNodeType();
+  output.WriteByte(static_cast<int8_t>(plan_type));
+
+  // Write database id and table id
+  if (!GetTable()) {
+    // The plan is not completed
+    return false;
+  }
+  oid_t database_id = GetTable()->GetDatabaseOid();
+  oid_t table_id = GetTable()->GetOid();
+
+  output.WriteInt(static_cast<int>(database_id));
+  output.WriteInt(static_cast<int>(table_id));
+
+  // If column has 0 item, just write the columnid_count with 0
+  int columnid_count = GetColumnIds().size();
+  output.WriteInt(columnid_count);
+
+  // If column has 0 item, nothing happens here
+  for (int it = 0; it < columnid_count; it++) {
+    oid_t col_id = GetColumnIds()[it];
+    output.WriteInt(static_cast<int>(col_id));
+  }
+
+  // Write predicate
+  if (GetPredicate() == nullptr) {
+    // Write the type
+    output.WriteByte(static_cast<int8_t>(EXPRESSION_TYPE_INVALID));
+  } else {
+    // Write the expression type
+    ExpressionType expr_type = GetPredicate()->GetExpressionType();
+    output.WriteByte(static_cast<int8_t>(expr_type));
 
     // Write predicate
-    if ( GetPredicate() == nullptr) {
-        // Write the type
-        output.WriteByte(static_cast<int8_t>(EXPRESSION_TYPE_INVALID));
-    } else {
-        // Write the expression type
-        ExpressionType expr_type = GetPredicate()->GetExpressionType();
-        output.WriteByte(static_cast<int8_t>(expr_type));
-
-        // Write predicate
-        GetPredicate()->SerializeTo(output);
-    }
-
-    // Write parent, but parent seems never be set or used right now
-    if (GetParent() == nullptr) {
-        // Write the type
-        output.WriteByte(static_cast<int8_t>(PLAN_NODE_TYPE_INVALID));
-    } else {
-        // Write the parent type
-        PlanNodeType parent_type = GetParent()->GetPlanNodeType();
-        output.WriteByte(static_cast<int8_t>(parent_type));
-
-        // Write parent
-        GetParent()->SerializeTo(output);
-    }
-
-    // Write the total length
-    int32_t sz = static_cast<int32_t>(output.Position() - start - sizeof(int));
-    PL_ASSERT(sz > 0);
-    output.WriteIntAt(start, sz);
-
-    return true;
+    GetPredicate()->SerializeTo(output);
+  }
+
+  // Write parent, but parent seems never be set or used right now
+  if (GetParent() == nullptr) {
+    // Write the type
+    output.WriteByte(static_cast<int8_t>(PLAN_NODE_TYPE_INVALID));
+  } else {
+    // Write the parent type
+    PlanNodeType parent_type = GetParent()->GetPlanNodeType();
+    output.WriteByte(static_cast<int8_t>(parent_type));
+
+    // Write parent
+    GetParent()->SerializeTo(output);
+  }
+
```

Comment:
Is this serialization code functionally correct? Presumably, you're serializing the tree top-down? If not, you should be. 

This code in its current form will result in an infinite recurrence since you go back up the three through the parent.

1. Do not serialize the parent. Serialization should be top-down.
2. You'll need to add a class that is a little smarter than the SerializeInput/SerializeOutput that can recognize nodes and expressions. Otherwise, every class will have a switch on node type and expression type for children and predicates.